From f73e5d05bc9d0693530c038e4a9b73e78f5c458a Mon Sep 17 00:00:00 2001
From: Hou Qi <qi.hou@nxp.com>
Date: Thu, 2 Nov 2023 14:34:11 +0900
Subject: [PATCH 14/19] V4L2VDA: Set OUTPUT format with parsed resolution for
 amphion

For VP8, VC1l, rv, Amphion needs to set correct resolution for OUTPUT
queue as it will be added to amphion customized header.

Upstream-Status: Inappropriate [NXP specific]
---
 .../v4l2/legacy/v4l2_video_decode_accelerator.cc    | 13 +++++++++++--
 .../gpu/v4l2/legacy/v4l2_video_decode_accelerator.h |  2 +-
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/media/gpu/v4l2/legacy/v4l2_video_decode_accelerator.cc b/media/gpu/v4l2/legacy/v4l2_video_decode_accelerator.cc
index e6edbcd97a896..261d53ffe217e 100644
--- a/media/gpu/v4l2/legacy/v4l2_video_decode_accelerator.cc
+++ b/media/gpu/v4l2/legacy/v4l2_video_decode_accelerator.cc
@@ -326,6 +326,9 @@ bool V4L2VideoDecodeAccelerator::CheckConfig(const Config& config) {
     return false;
   }
 
+  int width, height;
+  height = config.initial_expected_coded_size.height();
+  width = config.initial_expected_coded_size.width();
   // Capabilities check.
   struct v4l2_capability caps;
   unsigned int device_caps;
@@ -363,7 +366,7 @@ bool V4L2VideoDecodeAccelerator::CheckConfig(const Config& config) {
   if (!output_queue_)
     return false;
 
-  if (!SetupFormats())
+  if (!SetupFormats(width, height))
     return false;
 
   // We have confirmed that |config| is supported, tell the good news to the
@@ -2308,7 +2311,7 @@ bool V4L2VideoDecodeAccelerator::CreateInputBuffers() {
   return true;
 }
 
-bool V4L2VideoDecodeAccelerator::SetupFormats() {
+bool V4L2VideoDecodeAccelerator::SetupFormats(int width, int height) {
   DCHECK(decoder_thread_.task_runner()->BelongsToCurrentThread());
   DCHECK_EQ(decoder_state_, kInitialized);
   DCHECK(!input_queue_->IsStreaming());
@@ -2349,6 +2352,12 @@ bool V4L2VideoDecodeAccelerator::SetupFormats() {
     format.fmt.pix_mp.pixelformat = input_format_fourcc_;
     format.fmt.pix_mp.plane_fmt[0].sizeimage = input_size;
     format.fmt.pix_mp.num_planes = 1;
+    // For VP8, VC1l, rv, Amphion needs to set correct resolution for
+    // OUTPUT queue as it will be added to amphion customized header.
+    if(V4L2_PIX_FMT_VP8 == input_format_fourcc_){
+      format.fmt.pix_mp.width = width;
+      format.fmt.pix_mp.height = height;
+    }
   } else {
     format.type = V4L2_BUF_TYPE_VIDEO_OUTPUT;
     format.fmt.pix.pixelformat = input_format_fourcc_;
diff --git a/media/gpu/v4l2/legacy/v4l2_video_decode_accelerator.h b/media/gpu/v4l2/legacy/v4l2_video_decode_accelerator.h
index 89d830259add1..d2448871485f9 100644
--- a/media/gpu/v4l2/legacy/v4l2_video_decode_accelerator.h
+++ b/media/gpu/v4l2/legacy/v4l2_video_decode_accelerator.h
@@ -407,7 +407,7 @@ class MEDIA_GPU_EXPORT V4L2VideoDecodeAccelerator
   bool DestroyOutputBuffers();
 
   // Set input and output formats before starting decode.
-  bool SetupFormats();
+  bool SetupFormats(int, int);
   // Reset image processor and drop all processing frames.
   bool ResetImageProcessor();
 
-- 
2.34.1

